strawberry/permission.py

```diff
@@ -1,21 +1,185 @@
+
+    def handle_no_permission(self) -> None:
+        """
+        Default error raising for permissions.
+        This can be overridden to customize the behavior.
+        """
+
+        # Instantiate error class
+        error = self.error_class(self.message or "")
+
+        if self.error_extensions:
+            # Add our extensions to the error
+            if not error.extensions:
+                error.extensions = dict()
+            error.extensions.update(self.error_extensions)
+
+        raise error
+
+    @property
+    def schema_directive(self) -> object:
+        if not self._schema_directive:
+
+            class AutoDirective:
+                __strawberry_directive__ = StrawberrySchemaDirective(
+                    self.__class__.__name__,
+                    self.__class__.__name__,
+                    [Location.FIELD_DEFINITION],
+                    [],
+                )
+
+            self._schema_directive = AutoDirective()
+
+        return self._schema_directive
+
+
+class PermissionExtension(FieldExtension):
+    """
+    Handles permissions for a field
+    Instantiate this as a field extension with all of the permissions you want to apply
+
+    fail_silently: bool = False will return None or [] if the permission fails
+    instead of raising an exception. This is only valid for optional or list fields.
+
+    NOTE:
+    Currently, this is automatically added to the field, when using
+    field.permission_classes
+    This is deprecated behavior, please manually add the extension to field.extensions
+    """
+
+    def __init__(
+        self,
+        permissions: List[BasePermission],
+        use_directives: bool = True,
+        fail_silently: bool = False,
+    ):
+        self.permissions = permissions
+        self.fail_silently = fail_silently
+        self.return_empty_list = False
+        self.use_directives = use_directives
+
+    def apply(self, field: StrawberryField) -> None:
+        """
+        Applies all of the permission directives to the schema
+        and sets up silent permissions
+        """
+        if self.use_directives:
+            for permission in self.permissions:
+                if permission.schema_directive:
+                    field.directives.append(permission.schema_directive)
+        # We can only fail silently if the field is optional or a list
+        if self.fail_silently:
+            if isinstance(field.type, StrawberryOptional):
+                if isinstance(field.type.of_type, StrawberryList):
+                    self.return_empty_list = True
+            elif isinstance(field.type, StrawberryList):
+                self.return_empty_list = True
+            else:
+                raise PermissionFailSilentlyRequiresOptionalError(field)
+
+    def _handle_no_permission(self, permission: BasePermission) -> Any:
+        if self.fail_silently:
+            if self.return_empty_list:
+                return []
+            else:
+                return None
```

Comment:
nit: single return

(I know that there's a ruff check that would automatically do that change, but I'm not sure if it is enabled)

```suggestion
            return [] if self.return_empty_list else None
```